### PR TITLE
TESTS: Relax Assertion About Deleting Shard Dir

### LINKED ISSUE #32686: [CI] IndexRecoveryIT.testRerouteRecovery : Paths exist that should have been deleted
Build failures: https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+6.4+matrix-java-periodic/ES_BUILD_JAVA=java10,ES_RUNTIME_JAVA=java10,nodes=virtual&&linux/21/console

https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+matrix-java-periodic/ES_BUILD_JAVA=java10,ES_RUNTIME_JAVA=java11,nodes=virtual&&linux/224/console

Root cause:
```
com.carrotsearch.randomizedtesting.UncaughtExceptionError: Captured an uncaught exception in thread: Thread[id=5109, name=elasticsearch[node_t0][clusterApplierService#updateTask][T#1], state=RUNNABLE, group=TGRP-IndexRecoveryIT]
	at __randomizedtesting.SeedInfo.seed([E8CEC7E4E4750104:456574B0D12D6346]:0)
Caused by: java.lang.AssertionError: Paths exist that should have been deleted: [/var/lib/jenkins/workspace/elastic+elasticsearch+6.4+matrix-java-periodic/ES_BUILD_JAVA/java10/ES_RUNTIME_JAVA/java10/nodes/virtual&&linux/server/build/testrun/integTest/J1/temp/org.elasticsearch.indices.recovery.IndexRecoveryIT_E8CEC7E4E4750104-001/tempDir-004/data/nodes/0/indices/RJoRvWIVRDyM1toTfMV2tw/0]
	at __randomizedtesting.SeedInfo.seed([E8CEC7E4E4750104]:0)
	at org.elasticsearch.env.NodeEnvironment.assertPathsDoNotExist(NodeEnvironment.java:469)
	at org.elasticsearch.env.NodeEnvironment.deleteShardDirectoryUnderLock(NodeEnvironment.java:459)
	at org.elasticsearch.env.NodeEnvironment.deleteShardDirectorySafe(NodeEnvironment.java:399)
	at org.elasticsearch.indices.IndicesService.deleteShardStore(IndicesService.java:770)
	at org.elasticsearch.indices.store.IndicesStore$ShardActiveResponseHandler.lambda$allNodesResponded$2(IndicesStore.java:288)
	at org.elasticsearch.cluster.service.ClusterApplierService.lambda$runOnApplierThread$0(ClusterApplierService.java:306)
	at org.elasticsearch.cluster.service.ClusterApplierService$UpdateTask.apply(ClusterApplierService.java:155)
	at org.elasticsearch.cluster.service.ClusterApplierService.runTask(ClusterApplierService.java:399)
	at org.elasticsearch.cluster.service.ClusterApplierService$UpdateTask.run(ClusterApplierService.java:160)
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:624)
	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedEsThreadPoolExecutor.java:244)
	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedEsThreadPoolExecutor.java:207)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1135)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:844)
```

It does not reproduce:
```
REPRODUCE WITH: ./gradlew :server:integTest \
  -Dtests.seed=B4F936D0EF2B4D07 \
  -Dtests.class=org.elasticsearch.indices.recovery.IndexRecoveryIT \
  -Dtests.method="testRerouteRecovery" \
  -Dtests.security.manager=true \
  -Dtests.locale=fr-TN \
  -Dtests.timezone=Eire
```

### PR DESCRIPTION
**Pull request by original-brownbear**
* Allow empty state directory to prevent test from failing
* Closes #32686

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/env/NodeEnvironment.java [hunk 1]
```diff
@@ -485,13 +486,29 @@ public void deleteShardDirectoryUnderLock(ShardLock lock, IndexSettings indexSet
...
+    private static boolean assertPathsDoNotExist(final Path[] paths) throws IOException {
         Set<Path> existingPaths = new HashSet<>();
         for (Path path : paths) {
             if (FileSystemUtils.exists(path)) {
                 existingPaths.add(path);
             }
         }
+        // Relaxed assertion for the special case where only the empty state directory exists after deleting
+        // the shard directory because it was created again as a result of a metadata read action concurrently.
+        if (existingPaths.size() == 1) {
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/env/NodeEnvironment.java]: I wonder if it's easier to move the "check if there's only a left-over `MetaDataStateFormat.STATE_DIR_NAME`" logic into the `if (FileSystemUtils.exists(path)) {` condition above. Then you won't have to remove stuff again from `existingPaths`.

**Review by ywelsch — APPROVED** [review]
I've left one small suggestion, looks good otherwise. I think this should also go into the 6.4 branch as we are still actively running CI there.

**Comment by jasontedor** [comment]
> I think this should also go into the 6.4 branch as we are still actively running CI there.

And 5.6.

File: server/src/main/java/org/elasticsearch/env/NodeEnvironment.java (hunk 1 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/env/NodeEnvironment.java]: yea that would save a few lines :) sec adjusting

**Comment by original-brownbear** [comment]
@ywelsch applied your suggestion, much nicer this way I think ... with the initial version we would've also lost the logging of the name of the leftover paths.

**Review by ywelsch — APPROVED** [review]
LGTM

**Comment by original-brownbear** [comment]
@ywelsch thanks!

**Comment by DaveCTurner** [comment]
Backport to 5.6 wasn't totally trivial so I opened #34266.

**Pull request merged**